core/src/main/java/org/infinispan/interceptors/distribution/BaseDistributionInterceptor.java

```diff
@@ -144,288 +141,181 @@ public final BasicInvocationStage visitClearCommand(InvocationContext ctx, Clear
       return invokeNext(ctx, command);
    }
 
-   protected final CompletableFuture<InternalCacheEntry> retrieveFromProperSource(Object key, InvocationContext ctx,
-         FlagAffectedCommand command, boolean isWrite) throws Exception {
-      return doRetrieveFromProperSource(key, null, -1, ctx, command, isWrite);
-   }
-
-   private CompletableFuture<InternalCacheEntry> doRetrieveFromProperSource(Object key, InternalCacheEntry value,
-         int lastTopologyId, InvocationContext ctx, FlagAffectedCommand command, boolean isWrite) {
-      if (value != null)
-         return CompletableFuture.completedFuture(value);
-
-      final CacheTopology cacheTopology = stateTransferManager.getCacheTopology();
-      final int currentTopologyId = cacheTopology.getTopologyId();
+   protected final CompletableFuture<Void> remoteGet(InvocationContext ctx, AbstractDataCommand command,
+                                                     boolean isWrite) throws Exception {
+      final CacheTopology cacheTopology = checkTopologyId(command);
```

Comment:
> See how complex is the reasoning who must finish what before something else? And it makes that algorithm quite fragile.

True, this depends very closely on the way a rebalance is started/finished. But I'd argue you need to learn how rebalancing works anyway in order to test all the possible cases. Testing all the possible ways in which in which write commands can be retried because of a topology change is pretty painful, too, I'm not sure we are testing that with all the possible interceptors even now. And we still need new tests for write commands that are retried because of a remote get, especially in a transaction.

> sending readCH in OTR is not that expensive.

It can be expensive if the cluster is big, and that's exactly when we're most likely to get a delay in the topology installation.

> If TS > TR, we deal with the case that the owner lost some keys in TR that were in TS. 

You mean the owner gained some keys in TS that were not in TR, right?

> Rather than block and wait if it eventually gets the keys back, it could just return OTR without any CH, because it knows that its own CH is already outdated. If you say that the sender will always get some successful case, we can leave it up to that because it's easy to detect and throw e.g. RpcException if the assumption changes.

I'm saying the sender will always get some successful case *if* we allow the receiver to return an entry for which it is a write-owner, knowing that the new topology in which it is a read-owner was already sent out by the coordinator. If the receiver returns an OTR whenever it's not a read-owner, the sender is no longer guaranteed to get a valid response and it can't wait for a newer topology either (because it already has the newer topology).